serf/serf_test.go

```diff
@@ -2999,3 +2999,68 @@ func TestSerf_ValidateNodeName(t *testing.T) {
 	}
 
 }
+
+type reconnectOverride struct {
+	timeout time.Duration
+	called  bool
+}
+
+func (r *reconnectOverride) ReconnectTimeout(_ *Member, _ time.Duration) time.Duration {
+	r.called = true
+	return r.timeout
+}
+
+func TestSerf_perNodeReconnectTimeout(t *testing.T) {
+	ip1, returnFn1 := testutil.TakeIP()
+	defer returnFn1()
+
+	ip2, returnFn2 := testutil.TakeIP()
+	defer returnFn2()
+
+	override := reconnectOverride{timeout: 1 * time.Microsecond}
+
+	// Create the s1 config with an event channel so we can listen
+	eventCh := make(chan Event, 4)
+	s1Config := testConfig(t, ip1)
+	s1Config.ReconnectTimeout = 30 * time.Second
+	s1Config.ReconnectTimeoutOverride = &override
+	s1Config.EventCh = eventCh
+
+	s2Config := testConfig(t, ip2)
+
+	s1, err := Create(s1Config)
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+	defer s1.Shutdown()
+
+	s2, err := Create(s2Config)
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+	defer s2.Shutdown()
+
+	waitUntilNumNodes(t, 1, s1, s2)
+
+	_, err = s1.Join([]string{s2Config.NodeName + "/" + s2Config.MemberlistConfig.BindAddr}, false)
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	waitUntilNumNodes(t, 2, s1, s2)
+
+	err = s2.Shutdown()
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	waitUntilNumNodes(t, 1, s1)
+
+	// Since s2 shutdown, we check the events to make sure we got failures.
+	testEvents(t, eventCh, s2Config.NodeName,
+		[]EventType{EventMemberJoin, EventMemberFailed, EventMemberReap})
+
+	if !override.called {
+		t.Fatalf("The reconnect override was not used")
+	}
+}
```

Comment:
Should you add a reap test that uses artificially _inflated_ timeouts for a node so you can prove that some nodes get reaped as you'd expect and others linger as per the inputs? That might be easier to assert without awful sleeps slowing down the test than doing it the other way around.

Reply:
I actually think inflating the timeouts would make the test take longer. You still have to wait for memberlist to do its business of putting it into a suspect state then doing fallback pings etc. Then we would need to have an arbitrary sleep for a second or 2 to ensure that the node doesn't get reaped.

The test as written executes in about 2s with 1s between when we initiate the shutdown of the node to when we detect that its gone. To me having this test run for 2s seems fine and there are no arbitrary sleeps (just one retry.Run loop for a max of 7s but that should only execute for at most 1s until the reap happens - unless something breaks functionality and causes it not to be reaped)